File: frontend/src/shared/api/master-requests.ts
import { apiRequest } from './http';

export type MasterRequestStatus =
  | 'new'
  | 'accepted'
  | 'on_the_way'
  | 'in_progress'
  | 'completed'
  | 'handed_over'
  | 'cancelled';

export type MasterRequestComment = {
  id: string;
  master_request_id: string;
  author_id: string;
  author_name: string;
  body: string;
  created_at: string;
};

export type MasterRequestStatusLog = {
  id: string;
  master_request_id: string;
  changed_by_id: string | null;
  changed_by_name: string | null;
  from_status: MasterRequestStatus | null;
  to_status: MasterRequestStatus;
  note: string | null;
  created_at: string;
};

export type MasterRequestAttachment = {
  id: string;
  master_request_id: string;
  uploaded_by_id: string;
  uploaded_by_name: string;
  attachment_type: string;
  file_path: string;
  file_url: string;
  filename: string;
  mime_type: string | null;
  size_bytes: number | null;
  comment: string | null;
  created_at: string;
};

export type MasterRequestRecord = {
  id: string;
  branch_id: string;
  branch_name: string;
  requester_id: string;
  requester_name: string;
  assignee_id: string | null;
  assignee_name: string | null;
  title: string;
  description: string | null;
  address: string | null;
  client_name: string | null;
  client_phone: string | null;
  estimated_amount: string | number | null;
  final_amount: string | number | null;
  currency: string;
  status: MasterRequestStatus;
  geo_tracking_enabled: boolean;
  requested_at: string | null;
  accepted_at: string | null;
  started_at: string | null;
  completed_at: string | null;
  handed_over_at: string | null;
  cancelled_at: string | null;
  last_known_latitude: number | null;
  last_known_longitude: number | null;
  last_known_at: string | null;
  comments: MasterRequestComment[];
  status_logs: MasterRequestStatusLog[];
  attachments: MasterRequestAttachment[];
  geo_ping_count: number;
  created_at: string;
  updated_at: string;
};

export type MasterRequestStatusChangePayload = {
  status: MasterRequestStatus;
  note?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  captured_at?: string | null;
};

export async function fetchMasterRequests(accessToken: string, status?: MasterRequestStatus) {
  const search = status ? `?status=${status}` : '';
  return apiRequest<MasterRequestRecord[]>(`/master-requests${search}`, {
    method: 'GET',
    accessToken,
  });
}

export async function fetchMasterRequest(accessToken: string, id: string) {
  return apiRequest<MasterRequestRecord>(`/master-requests/${id}`, {
    method: 'GET',
    accessToken,
  });
}

export async function changeMasterRequestStatus(
  accessToken: string,
  id: string,
  payload: MasterRequestStatusChangePayload,
) {
  return apiRequest<MasterRequestRecord>(`/master-requests/${id}/status`, {
    method: 'POST',
    accessToken,
    body: JSON.stringify(payload),
  });
}

export async function addMasterRequestComment(accessToken: string, id: string, body: string) {
  return apiRequest<MasterRequestComment>(`/master-requests/${id}/comments`, {
    method: 'POST',
    accessToken,
    body: JSON.stringify({ body }),
  });
}

export async function uploadMasterRequestAttachment(
  accessToken: string,
  id: string,
  file: File,
  attachmentType: string = 'bso',
  comment?: string,
) {
  const formData = new FormData();
  formData.set('file', file);
  formData.set('attachment_type', attachmentType);
  if (comment) {
    formData.set('comment', comment);
  }

  return apiRequest<MasterRequestAttachment>(`/master-requests/${id}/attachments`, {
    method: 'POST',
    accessToken,
    body: formData,
  });
}

File: src/promouters/schemas/master_requests.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from promouters.models.enums import AttachmentType, MasterRequestStatus


class MasterRequestCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    address: str | None = Field(default=None, max_length=255)
    client_name: str | None = Field(default=None, max_length=255)
    client_phone: str | None = Field(default=None, max_length=64)
    estimated_amount: Decimal | None = Field(default=None, ge=0, decimal_places=2, max_digits=12)
    currency: str = Field(default="RUB", min_length=3, max_length=3)
    branch_id: UUID
    assignee_id: UUID | None = None
    requested_at: datetime | None = None


class MasterRequestUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    address: str | None = Field(default=None, max_length=255)
    client_name: str | None = Field(default=None, max_length=255)
    client_phone: str | None = Field(default=None, max_length=64)
    estimated_amount: Decimal | None = Field(default=None, ge=0, decimal_places=2, max_digits=12)
    final_amount: Decimal | None = Field(default=None, ge=0, decimal_places=2, max_digits=12)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    assignee_id: UUID | None = None


class MasterRequestStatusChange(BaseModel):
    status: MasterRequestStatus
    note: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    captured_at: datetime | None = None


class MasterRequestCommentCreate(BaseModel):
    body: str = Field(min_length=1)


class MasterRequestCommentRead(BaseModel):
    id: UUID
    master_request_id: UUID
    author_id: UUID
    author_name: str
    body: str
    created_at: datetime


class MasterRequestStatusLogRead(BaseModel):
    id: UUID
    master_request_id: UUID
    changed_by_id: UUID | None
    changed_by_name: str | None
    from_status: str | None
    to_status: str
    note: str | None
    created_at: datetime


class MasterGeoPingRead(BaseModel):
    id: UUID
    master_request_id: UUID
    master_id: UUID
    captured_at: datetime
    latitude: float
    longitude: float
    accuracy_meters: float | None
    status_at_capture: str


class MasterGeoPingCreate(BaseModel):
    captured_at: datetime
    latitude: float
    longitude: float
    accuracy_meters: float | None = None


class BSOAttachmentRead(BaseModel):
    id: UUID
    master_request_id: UUID
    uploaded_by_id: UUID
    uploaded_by_name: str
    attachment_type: str
    file_path: str
    file_url: str
    filename: str
    mime_type: str | None
    size_bytes: int | None
    comment: str | None
    created_at: datetime


class MasterRequestRead(BaseModel):
    id: UUID
    branch_id: UUID
    branch_name: str
    requester_id: UUID
    requester_name: str
    assignee_id: UUID | None
    assignee_name: str | None
    title: str
    description: str | None
    address: str | None
    client_name: str | None
    client_phone: str | None
    estimated_amount: Decimal | None
    final_amount: Decimal | None
    currency: str
    status: str
    geo_tracking_enabled: bool
    requested_at: datetime | None
    accepted_at: datetime | None
    started_at: datetime | None
    completed_at: datetime | None
    handed_over_at: datetime | None
    cancelled_at: datetime | None
    last_known_latitude: float | None
    last_known_longitude: float | None
    last_known_at: datetime | None
    comments: list[MasterRequestCommentRead]
    status_logs: list[MasterRequestStatusLogRead]
    attachments: list[BSOAttachmentRead]
    geo_ping_count: int
    created_at: datetime
    updated_at: datetime


class MasterRequestDeleteResponse(BaseModel):
    id: UUID
    message: str


class BSOAttachmentCreateMeta(BaseModel):
    attachment_type: AttachmentType = AttachmentType.BSO
    comment: str | None = None
